crates/bevy_transform/src/systems.rs

```diff
@@ -1,82 +1,99 @@
 use crate::components::{GlobalTransform, Transform};
-use bevy_ecs::prelude::{Changed, Entity, Query, With, Without};
+use bevy_ecs::{
+    entity::Entity,
+    query::{Changed, Or, With, Without},
+    system::{Local, Query},
+};
 use bevy_hierarchy::{Children, Parent};
 
+pub(crate) struct Pending {
+    parent: *const GlobalTransform,
+    changed: bool,
+    child: Entity,
+}
+
+// SAFE: Access to the parent pointer is only usable in this module, the values
+// are cleared after every system execution, and the system only uses one
+// thread. There is no way to move this type across multiple threads.
+unsafe impl Send for Pending {}
+// SAFE: Access to the parent pointer is only usable in this module, the values
+// are cleared after every system execution, and the system only uses one
+// thread. There is no way to access this type across multiple threads.
+unsafe impl Sync for Pending {}
+
+/// Update [`GlobalTransform`] component of entities without children or parents
+/// based on the [`Transform`] component.
+pub(crate) fn transform_propagate_flat_system(
+    mut root_query: Query<
+        (&Transform, &mut GlobalTransform),
+        (Without<Parent>, Without<Children>, Changed<Transform>),
+    >,
+) {
+    root_query.for_each_mut(|(transform, mut global_transform)| {
+        *global_transform = GlobalTransform::from(*transform);
+    });
+}
+
 /// Update [`GlobalTransform`] component of entities based on entity hierarchy and
 /// [`Transform`] component.
-pub fn transform_propagate_system(
+pub(crate) fn transform_propagate_system(
     mut root_query: Query<
         (
-            Option<(&Children, Changed<Children>)>,
             &Transform,
-            Changed<Transform>,
+            Or<(Changed<Transform>, Changed<Children>)>,
+            &Children,
             &mut GlobalTransform,
         ),
         Without<Parent>,
     >,
     mut transform_query: Query<
-        (&Transform, Changed<Transform>, &mut GlobalTransform),
+        (
+            &Transform,
+            Or<(Changed<Transform>, Changed<Children>)>,
+            Option<&Children>,
+            &mut GlobalTransform,
+        ),
         With<Parent>,
     >,
-    children_query: Query<(&Children, Changed<Children>), (With<Parent>, With<GlobalTransform>)>,
+    // Stack space for the depth-first search of a given hierarchy. Used as a Local to
+    // avoid reallocating the stack space used here.
```

Comment:
Heap space?

Reply:
It's technically used as a stack, just not **the** stack.